frontend/src/Store/Reducers/Session.reducer.ts

```diff
@@ -14,7 +15,25 @@ export default function sessionReducers(
         cardSet: action.payload.cardSet,
         cards: action.payload.cards,
       };
+    case ACTION_TYPES.SET_CARD_RATING:
+      return updateCardRating(state, action.payload);
     default:
       return state;
   }
 }
+const updateCardRating = (
+  state: SessionState,
+  cardData: Partial<CardData>
+): SessionState => {
+  let newState: SessionState = { ...state };
+  let cardIndex = newState.cards.findIndex((card) => card.id === cardData.id);
+  newState.cards = [
+    ...(state.cards as Array<CardData>).slice(0, cardIndex),
+    {
+      ...state.cards[cardIndex],
+      rating: cardData.rating,
+    } as CardData,
+    ...state.cards.slice(cardIndex + 1),
+  ];
```

Comment:
```suggestion
  newState.cards = [
    ...(state.cards).slice(0, cardIndex),
    {
      ...state.cards[cardIndex],
      rating: cardData.rating,
    } as CardData,
    ...state.cards.slice(cardIndex + 1),
  ];
```

Since we updated SessionState.cards to be explicitly 'never undefined', we can get rid of this little type casting